controllers/etcdcluster_controller.go

```diff
@@ -41,6 +43,171 @@ type EtcdClusterReconciler struct {
+			return nil, nil
+		}
+		// Unexpected error, some other problem?
+		return nil, err
+	}
+	// We found it because we got no error
+	return secret, nil
+}
+
+// getClientSecret determines if the client secret exists. Error is returned if there is some problem communicating with
+// Kubernetes.
+func (r *EtcdClusterReconciler) getClientSecret(ctx context.Context, clusterName, namespace string) (*v1.Secret, error) {
+	secret := &v1.Secret{}
+	err := r.Get(ctx, clientSecretName(clusterName, namespace), secret)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// We got the expected error, which is that it's not found
+			return nil, nil
+		}
+		// Unexpected error, some other problem?
+		return nil, err
+	}
+	// We found it because we got no error
+	return secret, nil
+}
+
+func caSecretName(clusterName, namespace string) types.NamespacedName {
+	name := fmt.Sprintf("%s-ca", clusterName)
+
+	return types.NamespacedName{
+		Name:      name,
+		Namespace: namespace,
+	}
+}
+
+// createCaSecret create the etcd ca secret
+func (r *EtcdClusterReconciler) createCaSecret(ctx context.Context, cluster *etcdv1alpha1.EtcdCluster) (*v1.Secret, error) {
+	secret, err := caSecretForCluster(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := r.Create(ctx, secret); err != nil {
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+// createClientSecret create the etcd client secret
+func (r *EtcdClusterReconciler) createClientSecret(ctx context.Context, cluster *etcdv1alpha1.EtcdCluster, caCert *v1.Secret) (*v1.Secret, error) {
+
+	secret, err := clientCertForCluster(cluster, caCert.Data["tls.crt"], caCert.Data["tls.key"])
+	if err != nil {
+		return nil, err
+	}
+
+	if err := r.Create(ctx, secret); err != nil {
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+func caSecretForCluster(cluster *etcdv1alpha1.EtcdCluster) (*v1.Secret, error) {
+	name := caSecretName(cluster.Name, cluster.Namespace)
+
+	caCert, caKey, err := tls.IssueCA()
+	if err != nil {
+		return nil, err
+	}
+
+	data := make(map[string][]byte)
+	data["tls.crt"] = caCert
+	data["tls.key"] = caKey
+
+	return &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name.Name,
+			Namespace: name.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(cluster, etcdv1alpha1.GroupVersion.WithKind("EtcdCluster")),
+			},
+			Labels: map[string]string{
+				appLabel:     appName,
+				clusterLabel: cluster.Name,
+			},
+		},
+		Data: data,
+	}, nil
+}
+
+func clientSecretName(clusterName, namespace string) types.NamespacedName {
+	name := fmt.Sprintf("%s-client", clusterName)
+
+	return types.NamespacedName{
+		Name:      name,
+		Namespace: namespace,
+	}
+}
+
+func clientCertForCluster(cluster *etcdv1alpha1.EtcdCluster, caCert, caKey []byte) (*v1.Secret, error) {
+	name := clientSecretName(cluster.Name, cluster.Namespace)
+
+	clientCert, clientKey, err := tls.Issue([]string{name.Name}, caCert, caKey)
```

Comment:
AFAIU this creates a client cert with the DNS name `clusterName-client`.
Shouldn't there be a svc with the same name going along with it? How else would you connect to the cluster as a client and have a correct name verification?

Reply:
> @FlorinPeter @cheahjs Any updates on this? Anything I could maybe help you with?

Sorry for delaying this PR, there are to many topics on my desk, my plan was to find a couple of hours next week to go over the review.

Reply:
You might be think of the peer's client certificate which requires that the Subject Alternative Name (SAN) includes DNS hostnames and/or IPs that match that of the peer. That certificate is generated in the EtcdPeer reconciler over here: https://github.com/improbable-eng/etcd-cluster-operator/pull/195/files#diff-2e9c68c9fc75a29d6beb2fe742c18d4b97a48e76796ccea4da15e663c3cfe6a4R659-R681

This bit of code is generating the client certificate for etcd clients, and the Common Name (CN) is used for logging in as the username in the CN (in this case, hardcoded to be `Etcd Cluster Operator` as per the `Subject` field).

Reference: https://etcd.io/docs/v3.4/op-guide/security/